temperature_probe: probe thermal drift compensation

Arksine opened the pull request:
This adds a `temperature_probe` module that supports configuring a temperature sensor for use with inductive probes.  This module implements the smoothing algorithm present in heaters and provides a temperature drift calibration procedure.  The module is intended to be generic, it can be used with any kind of inductive probe, where the calibration logic is farmed out to the probe specific implementation.  Also included in this PR is an implementation for the `probe_eddy_current` module.

The calibration procedure works by using the bed to heat the probe and requesting a manual probe before each sample.  This is intended to filter out thermal expansion.  Samples are taken at a specified temperature interval, by default this is set to every 2C.  The best calibration is achieved by taking samples across the widest temperature range possible.   After a manual probe, the `probe_eddy_current` module will take samples every .5mm, up to 4mm.  These samples are used to create a series of 2nd degree polynomials at each height.

Compensation uses the current frequency and coil temperature the find the upper and lower polynomials, then performs linear interpolation to calculate the correct frequency at the target temperature (ie: the temperature of the coil during `PROBE_EDDY_CURENT_CALIBRATE`).

Attached is a plot of the "stock" height calculate vs the corrected height for a probe that has a good temperature calibration.  This plot was measured by using the bed to heat the probe and performing a probe every 30 seconds, after which the height reported by the probe is collected.  `PROBE_EDDY_CURRENT_CALIBRATE` was performed with the probe around 35C.

![eddy-mounted-drift-correction](https://github.com/Klipper3d/klipper/assets/9563098/1c57b61c-7724-455c-9443-8e0e2492a489)

For comparison, I also collected data at a 1kHz sample rate.  In addition, this data was collected with `PROBE_EDDY_CURRENT_CALIBRATE` close to 60C:

![eddy-mounted-drift-correction-1khzrate](https://github.com/Klipper3d/klipper/assets/9563098/7262dc59-7be3-4a8e-b8d5-d60b7cfd2d11)

What stands out most in the above data is that the linear regression of the z-height is closer to the `z_offset` of `.5`.



This PR is built on #6558, so  we likely need to address any outstanding issues with that PR first.  Specific commits for this PR start at 108985cdc8af6c7cd315120b19b9fa06a9af2351.  The last two commits are minor fixes to `probe_eddy_current`, one fixing a stale attribute name and one slightly bumping the delay time after a probe to avoid "Unable to obtain probe_eddy_current sensor readings" errors.  If necessary I can drop these commits from this PR.

Arksine commented:
For those wondering about the status of this branch, I'm currently working on rebasing and resolving conflicts against the master branch.  My goal is to push an updated version this weekend (by June 22).

smwoodward commented:
> For those wondering about the status of this branch, I'm currently working on rebasing and resolving conflicts against the master branch. My goal is to push an updated version this weekend (by June 22).

Are we able to move off the BTT branch of klipper?

Arksine commented:
I have updated this branch to bring it in sync with the current master branch.  It should not be necessary to perform a new `PROBE_DRIFT_CALIBRATE` after updating to the newest implementation, however it will be necessary to perform a new `PROBE_EDDY_CURRENT_CALIBRATE`.

Another change to be aware of is that there is no longer a `SCAN_MODE` parameter in `BED_MESH_CALIBRATE`.  Instead there are two new options for the `METHOD` parameter, `scan` and `rapid_scan`.

bigtreetech commented on docs/Config_Reference.md:
```diff
@@ -2401,6 +2405,65 @@ temperature sensors that are reported via the M105 command.
+#   taken.  When the calibration procedure is complete the bed
+#   temperature will be set to zero.  When omitted the default
+#   behavior is not to set the bed temperature.
+#calibratation_extruder_temp:
```
should be `calibration_extruder_temp`

KevinOConnor commented:
Thanks for working on this.  Sorry for the delay in responding.

In general, this seems like useful functionality.  I have some high-level questions that would help me finish reviewing.

It looks like the temperature compensation algorithm works by creating 9 z mappings at z=0.05 to z=4.55 (0.500mm between each mapping).  Each of these Z mappings can translate a temperature to the expected ldc1612 sensor frequency at the given Z height.  The mappings are implemented using a quadratic formula (`expected_freq = c*temp*temp + b*temp + a`).  Each mapping is obtained by finding the least squares solution for a, b, c given all the (temperature, freq) pairs measured at the given Z height.  During runtime, the code compensates for temperature by adjusting a measured frequency - it finds the appropriate Z bucket given the current temperature and measured frequency, finds the expected frequency range for that Z bucket if the temperature was the nominal calibration temperature, and then interpolates the measured frequency back to the expected frequency at nominal temperature given those ranges.  Did I understand this system correctly?

What is the expected user workflow for this system?  Are users expected to run PROBE_DRIFT_CALIBRATE on each printer, or is it expected that most users will populate the temperature settings from some upstream source?  If the latter, do you have examples of tuned calibration settings?

In some of my previous tests, I got the impression that the ldc1612 was sensitive to both coil temperature and bed temperature.  Is it expected that: a) bed temperature isn't a big contributor to drift, b) bed temperature does contribute to drift but it is easy for users to stabilize the bed temperature prior to probing, c) bed temperature does contribute to drift but accounting for coil temperature is still better than nothing, or d) something else?

How would a typical user utilize SET_PROBE_DRIFT_ADJ_FACTOR?

Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement?  My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions.  I feel I may be missing something subtle.

Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today?  Has there been user feedback on calibration issues, accuracy tests, or similar?

Thanks again!
-Kevin

smwoodward commented:
> Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> 
> Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?

I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.

It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.

Sineos commented:
It seems the Beacon and Cartographer are using different approaches (or I have understood them wrong)

**Basic Calibration:**
* [Beacon](https://docs.beacon3d.com/quickstart/#6-calibrate-beacon)
* [Cartographer ](https://docs.cartographer3d.com/cartographer-probe/installation-and-setup/v2-and-v3-probes-w-input-shaping/cartographer-with-input-shaper-v2-and-v3-hybrid)

Both are using the paper test and only at one temperature.


**Extended Calibration:**
* Beacon: n/a
* [Cartographer ](https://docs.cartographer3d.com/cartographer-probe/fine-tuning/temperature-differential-calibration-beta)

In Cartographer, an automated process takes various frequency samples at different temperatures and computes profiles from them. No manual paper method is required.

For Beacon, I did not find such a procedure. It seems they rely on a factory-built calibration profile.

I cannot assess the accuracy of either product, but Cartographer appears to have the advantage of not requiring the paper method. 
It is likely that applying the manual paper method multiple times in a row will result in a significant measurement error due to the inherent limitations of the human user, who is unable to perform the paper method with the precision and accuracy required.

Arksine commented:
Hi Kevin, thanks for the review.

> It looks like the temperature compensation algorithm works by creating 9 z mappings at z=0.05 to z=4.55 (0.500mm between each mapping). Each of these Z mappings can translate a temperature to the expected ldc1612 sensor frequency at the given Z height. The mappings are implemented using a quadratic formula (expected_freq = c*temp*temp + b*temp + a). Each mapping is obtained by finding the least squares solution for a, b, c given all the (temperature, freq) pairs measured at the given Z height. During runtime, the code compensates for temperature by adjusting a measured frequency - it finds the appropriate Z bucket given the current temperature and measured frequency, finds the expected frequency range for that Z bucket if the temperature was the nominal calibration temperature, and then interpolates the measured frequency back to the expected frequency at nominal temperature given those ranges. Did I understand this system correctly?

This is correct.

> What is the expected user workflow for this system? Are users expected to run PROBE_DRIFT_CALIBRATE on each printer, or is it expected that most users will populate the temperature settings from some upstream source? If the latter, do you have examples of tuned calibration settings?

Its expected that the user run `PROBE_DRIFT_CALIBRATE` for each printer.  I compared data from multiple BTT Eddy probes and drift was not consistent between them, so I don't believe it is going to be possible to provide a factory preset.  That said, it may be possible that future iterations of the Eddy or other probe designs based on the ldc1612 are able to deliver a preset.

> In some of my previous tests, I got the impression that the ldc1612 was sensitive to both coil temperature and bed temperature. Is it expected that: a) bed temperature isn't a big contributor to drift, b) bed temperature does contribute to drift but it is easy for users to stabilize the bed temperature prior to probing, c) bed temperature does contribute to drift but accounting for coil temperature is still better than nothing, or d) something else?

In my testing bed temperature was not a significant factor with regard drift.  Coil temperature and distance from the bed were the primary factors.  For example, below are two plots captured with the bed at 85C AND 110C respectively.   This data was captured with the bed and chamber preheated, using a printed mount that decoupled the probe from the gantry.  The mount put the coil approximately ~3mm above the bed.

85C

![eddy-drift-85c-base-height](https://github.com/Klipper3d/klipper/assets/9563098/0524d7fa-569d-49a7-9539-fda3537b9bf2)


110C

![eddy-drift-110c-base-height](https://github.com/Klipper3d/klipper/assets/9563098/be2d88a4-b0e5-43cf-947a-629711c2429a)

> How would a typical user utilize SET_PROBE_DRIFT_ADJ_FACTOR?

I used it to disable thermal compensation by setting a factor of 0 during testing.  As I hinted mentioned, it may be possible for some probe designs to deliver a stock calibration and tweak the result by adjusting the factor.   I need to do more testing, but it might also be useful for tweaking calibration when switching build plates.

That said, I 100% understand if you want to remove this functionality until we receive confirmation with regard to its usage.
 
> Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.

A manual probe is necessary to remove the influence of thermal expansion from the frame and bed during the probe warming procedure.  A good calibration requires that we collect samples across the widest coil temperature range possible, so unfortunately we can't remove this influence by preheating the printer.  In the future, if its possible to use the ldc1612 to perform contact based probing, this operation could be automated. 

> Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?

BTT has users running this code based on the version of `probe_eddy_current` prior to the recent refactor.  General feedback I have received is that it works well when calibrated correctly, however the calibration procedure is difficult.

smwoodward commented:
@KevinOConnor Something that I wanted to as or run by you with these new probes, I did install on of the eddy's on my ratrig and I kept the Pinda probe on also, and I moved it to be a z-endstop which works for homing, but is there a way to add the capability to do things like homing and QGL/z-tilt with one probe and do the mesh with the different probe?

KevinOConnor commented:
> In my testing bed temperature was not a significant factor with regard drift.

Thanks for the responses!

Just to further clarify the impact of bed temperature - the two graphs you provided don't look that similar to me.  For example, at 85c bed temperature the probe at 60c seems to report 3.270Mhz while at 110c bed temperature the probe at 60c seems to report 3.269Mhz.  Granted that's small relative to the coil swing (1khz for 25c bed temperature change relative to about 5Khz for 25c probe temperature change).  I'm not sure it's small relative to our probe Z height accuracy goals though.  I think I may be missing something.

-Kevin

KevinOConnor commented:
@smwoodward - for questions on using Klipper please follow the contact directions at: https://www.klipper3d.org/Contact.html

-Kevin

KevinOConnor commented:
Thanks for working on this.  In general it seems worthwhile to add to Klipper.

My main comment is about the interface between module components.  What do you think about an alternate interface between temperature_probe.py and probe_eddy_current.py?  This PR changes probe_eddy_current to detect and register itself with the temperature_probe code (`EddyDriftCompensation:self.temp_sensor.register_calibration_helper(self)`).  Could we flip that around and have the temperature_probe code lookup probe_eddy_current and call a new `probe_eddy_current.register_temperature_compensation(self)`?

The reason I think that may be useful is that it seems the probe temperature calibration interface (TemperatureProbe to EddyDriftCompensation) is more complex than the frequency adjustment interface (EddyDriftCompensation to EddyCalibration).  Thus if we can move EddyDriftCompensation into temperature_probe.py then I think the code interfaces might be more apparent.

I think we could also move `calibration_temp` to EddyDriftCompensation.  The EddyCalibration code could call a `self.drift_comp.note_start_z_calibration()` at the start of calibration and call a `self.drift_comp.note_finish_z_calibration()` at the end of calibration (which could signal to EddyDriftCompensation to store the `calibration_temp` in the temperature_probe config section during auto save).  Thus EddyCalibration would have only a few temperature specific methods (`def register_temperature_compensation()`, `self.drift_comp.note_start_z_calibration()`, `self.drift_comp.note_finish_z_calibration()`, `self.drift_comp.adjust_freqs()`, `self.drift_comp.unadjust_freq()`).

I have a few other code comments which I'll follow up with below.

Cheers,
-Kevin

KevinOConnor commented on klippy/extras/probe_eddy_current.py:
```diff
@@ -38,7 +41,8 @@ def load_calibration(self, cal):
     def apply_calibration(self, samples):
         for i, (samp_time, freq, dummy_z) in enumerate(samples):
-            pos = bisect.bisect(self.cal_freqs, freq)
+            adj_freq = self.drift_comp.adjust_freq(freq, self.cal_temp)
+            pos = bisect.bisect(self.cal_freqs, adj_freq)
```
This looks like it could cause a subtle race condition.  The `apply_calibration()` call works on a block of frequency measurements.  However, `adjust_freq()` could pull a new temperature on each individual call (as `TemperatureProbe:_temp_callback()` runs in a background thread).  That could mean that in some subtle situations a different temperature could be applied midway through a set of samples.  Probably better to consistently use a single temperature for each block of samples.

One way to do this would be to extract the temperature outside of the loop and then pass that temperature to each `adjust_freq()` call.  Another option (that might be simpler and more efficient) could be to pass all the frequencies as a list to a new `adjust_freqs()` call.

Arksine replied:
One thing I considered was moving the call to `adjust_freq()` from `apply_calibration` to `EddyCalibration.freq_to_height()`.  This would reduce overhead from performing the adjustment on all samples, with the downside that the adjustment would't be applied to the z values in the dump command.

That said, I'm fine with capturing the probe temperature outside of the loop and applying it to all samples.

KevinOConnor replied:
I think it would be helpful if the API server continued to export a Z value that is temperature adjusted.

KevinOConnor commented on klippy/mathutil.py:
```diff
@@ -135,3 +135,76 @@ def matrix_sub(m1, m2):
+    afh = m[0][0] * m[2][1] * m[1][2]
+    return aei + bfg + cdh - ceg - bdi - afh
+
+class Polynomial2d:
+    def __init__(self, a, b, c):
+        self.a = a
+        self.b = b
+        self.c = c
+
+    def __call__(self, xval):
+        return self.c * xval * xval + self.b * xval + self.a
+
+    def get_coefs(self):
+        return (self.a, self.b, self.c)
+
+    def __str__(self):
+        return "%f, %f, %f" % (self.a, self.b, self.c)
```
I don't think `Polynomial2d` is a good fit for mathutil.py.  We can certainly add the linear squares code and/or the matrix determinant code to mathutil.py.  (Interestingly, I think @dmbutyugin recently made a similar implementation for his generic_cartesian code.)

The `EddyDriftCompensation` code has some subtle dependencies on `Polynomianl2d` that I'm unsure would be apparent to developers looking at just mathutil.py (eg, the `str()` format becomes a user-facing api as it is written directly to the config file).

One simple solution would be to relocate `Polynomial2d` to temperature_probe.py .

Arksine replied:
Sounds good.

dmbutyugin replied:
Yes, the code can be found here
https://github.com/dmbutyugin/klipper/blob/5f4df09c99d9e92a3f9896c9e8107659d0efc0d0/klippy/mathutil.py#L143-L152
and the calculation of position using pseudoinverse:
https://github.com/dmbutyugin/klipper/blob/generic-cartesian/klippy/kinematics/generic_cartesian.py#L124-L128
which is also an least squares problem solution. I think this one is a bit less verbose, so feel free to use, but up to you.

KevinOConnor commented on klippy/extras/probe_eddy_current.py:
```diff
@@ -436,5 +446,252 @@ def start_probe_session(self, gcmd):
+            "file and restart the printer."
+            % (self.name, len(polynomials))
+        )
+        try:
+            # Dump collected data to temporary file
+            import json
+            ctime = int(self.printer.get_reactor().monotonic())
+            tmpfname = "/tmp/eddy-probe-drift-%d.json" % (ctime)
+            out = {
+                "polynomial_coefs": [c.get_coefs() for c in polynomials],
+                "legend": ["temperature", "frequency"],
+                "data": cal_samples,
+                "start_z": .05,
+                "sample_z_dist": .5
+            }
+            with open(tmpfname, "w") as f:
+                f.write(json.dumps(out))
+        except Exception:
+            logging.exception("Failed to write %s" % (tmpfname))
```
I think we should avoid writing ad-hoc files from Klipper.  If this is just debug code, maybe we should remove it before merging in the main Klipper repo.  If it's generally useful information then I think we should look to export it via the Klipper API interface.

The ad-hoc files are hard for users to use, hard for us to document, and can become a burden to maintain as future devs wont necessarily know if they can change the contents.

Arksine replied:
I'll remove it.  Its purpose was primarily for collecting data from other users testing the module.

KevinOConnor commented on klippy/extras/temperature_probe.py:
```diff
@@ -0,0 +1,375 @@
+    def register_calibration_helper(self, helper):
+        self.cal_helper = helper
+
+    def _temp_callback(self, read_time, temp):
+        if not temp:
+            return
+        with self.temp_lock:
+            time_diff = read_time - self.last_temp_read_time
+            self.last_temp_read_time = read_time
+            temp_diff = temp - self.smoothed_temp
+            adj_time = min(time_diff * self.inv_smooth_time, 1.)
+            self.smoothed_temp += temp_diff * adj_time
+            self.measured_min = min(self.measured_min, self.smoothed_temp)
+            self.measured_max = max(self.measured_max, self.smoothed_temp)
+            st = self.smoothed_temp
+            if self.in_calibration and st >= self.next_auto_temp:
+                self.next_auto_temp = 99999999.
+                self.printer.get_reactor().register_async_callback(
+                    lambda e, gc=self.gcode: gc.run_script("PROBE_DRIFT_NEXT")
+                )
```
This looks like it could have subtle race conditions (as `_temp_callback()` runs in a background thread).

The `self.temp_lock` is only ever taken here, which doesn't seem correct.  (The lock only works if other places in the code also acquire the lock.)

Many of the variables (measured_min, measured_max, in_calibration, next_auto_temp) could change mid-execution (both here and in `get_status()` etc) due to the background thread.

Also, technically, it's possible for PROBE_DRIFT_NEXT to be queued from `_temp_callback()` at a similar moment that the user manually calls `PROBE_DRIFT_NEXT` - which could cause two back-to-back executions.

Arksine replied:
Ah, I must have had async programming on my mind.  FWIW, its probably not a great plan to acquire a thread lock on the main thread since acquisition can be a blocking operation.  I wonder if it would be a good idea to create a wrapper around the lock using the reactor, testing if the lock is held and pausing until it isn't.

KevinOConnor replied:
I'm not sure the reactor can help much as it only has limited thread safety itself (only `register_async_callback()` can be called from the background thread) and the background thread isn't running a reactor (only the main thread has it).

One possible alternative is to store groups of values in tuples, as Python then assures each assignment is atomic.  The clocksync.py code does this.

Something like (totally untested):
```python
    def _temp_callback(self, read_time, temp):
        time_diff = read_time - self.last_temp_read_time
        self.last_temp_read_time = read_time
        smoothed_temp, measured_min, measured_max = self.last_measurement
        temp_diff = temp - smoothed_temp
        adj_time = min(time_diff * self.inv_smooth_time, 1.)
        smoothed_temp += temp_diff * adj_time
        measured_min = min(measured_min, smoothed_temp)
        measured_max = max(measured_max, smoothed_temp)
        self.last_measurement = (smoothed_temp, measured_min, measured_max)
        if self.in_calibration and smoothed_temp >= self.next_auto_temp:
            self.printer.get_reactor().register_async_callback(
                self._check_kick_next)

    def _check_kick_next(self, eventtime):
        smoothed_temp, measured_min, measured_max = self.last_measurement
        if self.in_calibration and smoothed_temp >= self.next_auto_temp:
            self.next_auto_temp = 99999999.
            self.gcode.run_script("TEMPERATURE_PROVE_NEXT")
```

The idea here is that smoothed_temp, measured_min, measured_max are always atomically written as a group, and thus any readers will get those values as a group (never a mix from two possible `_temp_callback()` invocations).  Also, a "dirty read" of self.in_calibration and self.next_auto_temp isn't an issue as long as the values are rechecked in the main thread.

KevinOConnor commented on docs/G-Codes.md:
```diff
@@ -1415,3 +1422,34 @@ command will probe the points specified in the config and then make independent
+[temperature_probe config section](Config_Reference.md#temperature_probe)
+is enabled.
+
+#### PROBE_DRIFT_CALIBRATE
```
What about `TEMPERATURE_PROBE_CALIBRATION` instead of `PROBE_DRIFT_CALIBRATE`?  Putting the module name in the command helps reduce the chance of command conflicts in the future.

Arksine replied:
I'm good with that.

KevinOConnor commented on docs/G-Codes.md:
```diff
@@ -1006,6 +1006,13 @@ appropriate DRIVE_CURRENT for the sensor. After running this command
 use the SAVE_CONFIG command to store that new setting in the
 printer.cfg config file.
 
+#### SET_PROBE_DRIFT_ADJ_FACTOR
+`SET_PROBE_DRIFT_ADJ_FACTOR PROBE=<config_name> FACTOR=<new_factor>`:
+When thermal drift compensation is calibrated this command may be
+used to set the `drift_adj_factor` during runtime.  This value
+is a multiplier applied to the calculated drift compensation and may be
+used to fine tune the result.
```
For what it is worth, this command seems very subtle to me.  Might be better to hold off on it until we know it is needed.  Similar with the config option.

Or another alternative might be to have something like a `PROBE_TEMPERATURE_ENABLE` command to let users manually enable/disable temperature_probe without the subtle fine tuning.

Arksine replied:
Sounds good.

KevinOConnor commented on klippy/extras/temperature_probe.py:
```diff
@@ -0,0 +1,375 @@
+        # If configured move to heating position and turn on extruder
+        try:
+            self._move_to_start()
+        except Exception:
```
Very minor, but this should probably be `except self.printer.config_error` (as above).  An internal coding error probably shouldn't be caught here.

KevinOConnor reviewed (commented):
Thanks again - see my comments below.

-Kevin

bigtreetech commented:
> @KevinOConnor Something that I wanted to as or run by you with these new probes, I did install on of the eddy's on my ratrig and I kept the Pinda probe on also, and I moved it to be a z-endstop which works for homing, but is there a way to add the capability to do things like homing and QGL/z-tilt with one probe and do the mesh with the different probe?

This is already a standard part of how klipper works but as Kevin said, this is not the forum to discuss questions unrelated to the specifics of this PR which are dealing with thermal comp.

Arksine commented:
> Just to further clarify the impact of bed temperature - the two graphs you provided don't look that similar to me. For example, at 85c bed temperature the probe at 60c seems to report 3.270Mhz while at 110c bed temperature the probe at 60c seems to report 3.269Mhz. Granted that's small relative to the coil swing (1khz for 25c bed temperature change relative to about 5Khz for 25c probe temperature change). I'm not sure it's small relative to our probe Z height accuracy goals though. I think I may be missing something.

I can explain the discrepancy.  When collecting data it became immediately apparent that I needed to decouple the probe from the gantry to remove the influence of thermal expansion.  BTT designed a mount that encases the probe and places the coil ~3mm above the surface.  My collection procedure was to preheat the chamber, connect the probe and place it roughly in the center of the bed, collect data, then remove when done to allow the probe to cool down. 

Since the probe isn't fixed, small deviations can change the nominal frequency.  Conditions such as the PEI being slightly thicker in one area, imperfections in the printed mount, imperfections in the build plate, etc can all influence then nominal frequency.  Put differently, I could collect two data sets at 110C and they would show the same minor deviation.

> My main comment is about the interface between module components. What do you think about an alternate interface between temperature_probe.py and probe_eddy_current.py? This PR changes probe_eddy_current to detect and register itself with the temperature_probe code (EddyDriftCompensation:self.temp_sensor.register_calibration_helper(self)). Could we flip that around and have the temperature_probe code lookup probe_eddy_current and call a new probe_eddy_current.register_temperature_compensation(self)?

Sure, that wont be a problem.


> I think we could also move calibration_temp to EddyDriftCompensation. The EddyCalibration code could call a self.drift_comp.note_start_z_calibration() at the start of calibration and call a self.drift_comp.note_finish_z_calibration() at the end of calibration (which could signal to EddyDriftCompensation to store the calibration_temp in the temperature_probe config section during auto save). Thus EddyCalibration would have only a few temperature specific methods (def register_temperature_compensation(), self.drift_comp.note_start_z_calibration(), self.drift_comp.note_finish_z_calibration(), self.drift_comp.adjust_freqs(), self.drift_comp.unadjust_freq()).

Sound good.  I'll work on that change.  In addition, I'll address each review comment individually as I go along.

Arksine commented:
I have pushed some updates, the changes of which are outlined below:

- moved the `Polynomial2D` class from `mathutil.py` to `temperature_probe.py`
- moved the `EddyDriftCompensation` class from `probe_eddy_current.py` to `temperature_probe.py`.  The `probe_eddy_current` module now contains a "dummy" class serves as the default when no temperature probe is registered.
- Reversed registration logic.  The `temperature_probe` now registers the compensation helper with the underlying probe implementation by calling its `register_drift_compensation()` method.
- Moved calibration temperature tracking from the `probe_eddy_current.EddyCalibration` class to `temperature_probe.EddyDriftCompensation`.  Eddy Probe Calibration calls the suggested `note_z_calibration_start()` and `note_z_calibration_finish()` methods of the compensation helper.
- Use one temperature reading for each batch of ldc samples received when applying drift compensation.
- Remove the debug file output
- Remove the thread lock in `TempertureProbe`.  Use the suggested method of storing the smoothed, min, and max tempertures in a tuple for atomic writes.  Implement the suggested `_check_kick_next()` callback to avoid a potential race when calling "TEMPERATURE_PROBE_NEXT"
- Rename all gcode commands prefixed withe `PROBE_DRIFT_` to `TEMPERATURE_PROBE_`
- Remove the `SET_PROBE_DRIFT_ADJ_FACTOR` gcode command
- Implement `TEMPERATURE_PROBE_ENABLE` gcode command.  Report the current enabled state in `TemperatureProbe.get_status()`
- In the `TEMPERATURE_PROBE_CALIBRATE` gcode handler, catch `self.printer.command_error` when calling `_move_to_start()`.

I believe that covers everything.  This change moves all drift compensation related items in the autosave from the `probe_eddy_current` section to the `temperature probe` section.  Testers that don't wish to redo their calibration can manually update the section (generally it isn't recommended to modify the autosave, but it seems appropriate for this purpose).  For example, the autosave section should now contain a snippet that looks like the following.

```ini
#*# [temperature_probe btt_eddy]
#*# calibration_temp = 42.919721
#*# drift_calibration =
#*# 	3265957.923222, -161.557392, 0.448391
#*# 	3256524.613344, -106.776906, 0.145617
#*# 	3249162.938671, -68.684932, -0.075612
#*# 	3243413.959281, -44.068844, -0.206610
#*# 	3238864.481265, -26.708818, -0.301046
#*# 	3235263.558343, -15.621579, -0.355331
#*# 	3232252.151603, -4.210931, -0.421378
#*# 	3229739.335973, 7.083410, -0.497587
#*# 	3227831.886549, 9.409186, -0.497820
#*# drift_calibration_min_temp = 32.18233127179636
```

KevinOConnor commented:
Thanks.  It looks fine to me.  If there are no further comments I'll look to commit in a few days.  I guess I'll look to do a "squash and merge".

-Kevin

KevinOConnor commented on klippy/extras/temperature_probe.py:
```diff
@@ -0,0 +1,718 @@
+            )
+
+    def _temp_callback(self, read_time, temp):
+        if not temp:
+            return
```
FWIW, this test looks a bit odd to me as no sensor should be passing on a bogus zero value (if a zero value is relayed it would indicate zero Celsius).  If there are sensors passing on bad data we should definitely look to fix those sensors.

-Kevin

Arksine replied:
Ok, I'll remove it.  IIRC I included a similar check to the temperature callback in `temperature_sensor.py` when adding support for min/max temps. It was reported that some sensors would report an initial value of 0 which would impact the minimum stored value.  I agree that its better to debug the sensor reporting that value.

Arksine commented:
> Thanks. It looks fine to me. If there are no further comments I'll look to commit in a few days. I guess I'll look to do a "squash and merge".

Sounds good.  I just removed the `if temp` check, and did a fixup on the commits so you can rebase and merge if that is your preference.

KevinOConnor commented:
Thanks.  I committed this change.  I was just giving a few days to see if others had comments.

Cheers,
-Kevin

smwoodward commented:
> > > Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> > > Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?
> > 
> > 
> > I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.
> > It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.
> 
> I totally agree. the paper testing requires opening the chamber door, which causes rapid loss of chamber temperature and introduces rapid changes in coil temperature.

I don't think the thermal loss is that great to the coil temperature from what I saw, I can say that the thermal loss of the air in the chamber would extend the required time in heating up the coil as the chamber environment would need to heat back up to keep increasing the coil temp.  There would still be a good amount of radiation heat transfer because of the distance between the bed and coil, but the convection transfer would take time to regain the temperature once the chamber is closed again.

qqqwqk commented:
> > > > Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> > > > Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?
> > > 
> > > 
> > > I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.
> > > It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.
> > 
> > 
> > I totally agree. the paper testing requires opening the chamber door, which causes rapid loss of chamber temperature and introduces rapid changes in coil temperature.
> 
> I don't think the thermal loss is that great to the coil temperature from what I saw, I can say that the thermal loss of the air in the chamber would extend the required time in heating up the coil as the chamber environment would need to heat back up to keep increasing the coil temp. There would still be a good amount of radiation heat transfer because of the distance between the bed and coil, but the convection transfer would take time to regain the temperature once the chamber is closed again.

I'm just raising an idea: maybe thermal expansion of the frame and bed should not be taken into account when calibrating the probe, which would simplify the manual work of calibration(As far as I know: Not even Hannibal would try to cook his fingers). 
Thermal expansion should be considered as a separate item (perhaps we should consider using z-endstop to try to detect/calculate thermal expansion)

smwoodward commented:
> > > > > Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> > > > > Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?
> > > > 
> > > > 
> > > > I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.
> > > > It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.
> > > 
> > > 
> > > I totally agree. the paper testing requires opening the chamber door, which causes rapid loss of chamber temperature and introduces rapid changes in coil temperature.
> > 
> > 
> > I don't think the thermal loss is that great to the coil temperature from what I saw, I can say that the thermal loss of the air in the chamber would extend the required time in heating up the coil as the chamber environment would need to heat back up to keep increasing the coil temp. There would still be a good amount of radiation heat transfer because of the distance between the bed and coil, but the convection transfer would take time to regain the temperature once the chamber is closed again.
> 
> I'm just raising an idea: maybe thermal expansion of the frame and bed should not be taken into account when calibrating the probe, which would simplify the manual work of calibration(As far as I know: Not even Hannibal would try to cook his fingers). Thermal expansion should be considered as a separate item (perhaps we should consider using z-endstop to try to detect/calculate thermal expansion)

Perhaps someone else can speak more about exactly what is going on, but from my understanding the temperature calibration isn't more about thermal expansion as the readings the eddy current sensor change based on temperature.  It's a bit different than thermal expansion which would cause the physical parts to change shape.  It's more that everything else being equal (sensor being exactly the same distance away from the metal surface no matter the temperature) that if the sensor is reading 2.7 gigawatts (or what ever it is measuring) at 24*, that if the sensor alone goes to 35* it will read 3.7 gigawatts, but nothing else has changed except for the temp of the sensor.  Thermal expansion would cause the gap to physically change between the two items.

qqqwqk commented:
> > > > > > Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> > > > > > Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?
> > > > > 
> > > > > 
> > > > > I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.
> > > > > It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.
> > > > 
> > > > 
> > > > I totally agree. the paper testing requires opening the chamber door, which causes rapid loss of chamber temperature and introduces rapid changes in coil temperature.
> > > 
> > > 
> > > I don't think the thermal loss is that great to the coil temperature from what I saw, I can say that the thermal loss of the air in the chamber would extend the required time in heating up the coil as the chamber environment would need to heat back up to keep increasing the coil temp. There would still be a good amount of radiation heat transfer because of the distance between the bed and coil, but the convection transfer would take time to regain the temperature once the chamber is closed again.
> > 
> > 
> > I'm just raising an idea: maybe thermal expansion of the frame and bed should not be taken into account when calibrating the probe, which would simplify the manual work of calibration(As far as I know: Not even Hannibal would try to cook his fingers). Thermal expansion should be considered as a separate item (perhaps we should consider using z-endstop to try to detect/calculate thermal expansion)
> 
> Perhaps someone else can speak more about exactly what is going on, but from my understanding the temperature calibration isn't more about thermal expansion as the readings the eddy current sensor change based on temperature. It's a bit different than thermal expansion which would cause the physical parts to change shape. It's more that everything else being equal (sensor being exactly the same distance away from the metal surface no matter the temperature) that if the sensor is reading 2.7 gigawatts (or what ever it is measuring) at 24*, that if the sensor alone goes to 35* it will read 3.7 gigawatts, but nothing else has changed except for the temp of the sensor. Thermal expansion would cause the gap to physically change between the two items.

Heating the coil sensor will also heat the bed and frame, and during this process, the distance between the sensor and the bed cannot remain constant.

smwoodward commented:
> > > > > > > Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> 
> > > > > > > Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?
> 
> > > > > > 
> 
> > > > > > 
> 
> > > > > > I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.
> 
> > > > > > It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.
> 
> > > > > 
> 
> > > > > 
> 
> > > > > I totally agree. the paper testing requires opening the chamber door, which causes rapid loss of chamber temperature and introduces rapid changes in coil temperature.
> 
> > > > 
> 
> > > > 
> 
> > > > I don't think the thermal loss is that great to the coil temperature from what I saw, I can say that the thermal loss of the air in the chamber would extend the required time in heating up the coil as the chamber environment would need to heat back up to keep increasing the coil temp. There would still be a good amount of radiation heat transfer because of the distance between the bed and coil, but the convection transfer would take time to regain the temperature once the chamber is closed again.
> 
> > > 
> 
> > > 
> 
> > > I'm just raising an idea: maybe thermal expansion of the frame and bed should not be taken into account when calibrating the probe, which would simplify the manual work of calibration(As far as I know: Not even Hannibal would try to cook his fingers). Thermal expansion should be considered as a separate item (perhaps we should consider using z-endstop to try to detect/calculate thermal expansion)
> 
> > 
> 
> > Perhaps someone else can speak more about exactly what is going on, but from my understanding the temperature calibration isn't more about thermal expansion as the readings the eddy current sensor change based on temperature. It's a bit different than thermal expansion which would cause the physical parts to change shape. It's more that everything else being equal (sensor being exactly the same distance away from the metal surface no matter the temperature) that if the sensor is reading 2.7 gigawatts (or what ever it is measuring) at 24*, that if the sensor alone goes to 35* it will read 3.7 gigawatts, but nothing else has changed except for the temp of the sensor. Thermal expansion would cause the gap to physically change between the two items.
> 
> 
> 
> heating the coil sensors also heat the bed and frame simultaneously, and during this process the distance between the sensor and the bed does not remain constant.
> 
> 

That's why you do the paper test to make sure that the nozzle is still at the distance it should be, then it takes readings of the eddy currents at the exact same spot to know how they change. From my understanding the sensor readings change much more drastically due to temperature than thermal expansion affects the measurements.

qqqwqk commented:
> > > > > > > > Why does PROBE_DRIFT_CALIBRATE perform a manual probe at the start of each temperature measurement? My initial thought would be to perform the Z/freq/temp sampling at identical Z carriage positions. I feel I may be missing something subtle.
> > 
> > 
> > > > > > > > Do you know if this PROBE_DRIFT_CALIBRATE code is in wide use today? Has there been user feedback on calibration issues, accuracy tests, or similar?
> > 
> > 
> > > > > > >
> > 
> > 
> > > > > > >
> > 
> > 
> > > > > > > I just set this up recently off of their forked version, and what it did at each temperature measurement was to do a paper test with the nozzle, and when I clicked accept it moved and did measurements, then went back to the position and waited till the next temperature which I used the paper again to set the manual probe and it repeated.
> > 
> > 
> > > > > > > It wasn't difficult to do, but it would be nice if there was a different way of doing it than sitting by the printer for 30+ minutes to do the paper test over and over.
> > 
> > 
> > > > > >
> > 
> > 
> > > > > >
> > 
> > 
> > > > > > I totally agree. the paper testing requires opening the chamber door, which causes rapid loss of chamber temperature and introduces rapid changes in coil temperature.
> > 
> > 
> > > > >
> > 
> > 
> > > > >
> > 
> > 
> > > > > I don't think the thermal loss is that great to the coil temperature from what I saw, I can say that the thermal loss of the air in the chamber would extend the required time in heating up the coil as the chamber environment would need to heat back up to keep increasing the coil temp. There would still be a good amount of radiation heat transfer because of the distance between the bed and coil, but the convection transfer would take time to regain the temperature once the chamber is closed again.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > I'm just raising an idea: maybe thermal expansion of the frame and bed should not be taken into account when calibrating the probe, which would simplify the manual work of calibration(As far as I know: Not even Hannibal would try to cook his fingers). Thermal expansion should be considered as a separate item (perhaps we should consider using z-endstop to try to detect/calculate thermal expansion)
> > 
> > 
> > > 
> > 
> > 
> > > Perhaps someone else can speak more about exactly what is going on, but from my understanding the temperature calibration isn't more about thermal expansion as the readings the eddy current sensor change based on temperature. It's a bit different than thermal expansion which would cause the physical parts to change shape. It's more that everything else being equal (sensor being exactly the same distance away from the metal surface no matter the temperature) that if the sensor is reading 2.7 gigawatts (or what ever it is measuring) at 24*, that if the sensor alone goes to 35* it will read 3.7 gigawatts, but nothing else has changed except for the temp of the sensor. Thermal expansion would cause the gap to physically change between the two items.
> > 
> > 
> > heating the coil sensors also heat the bed and frame simultaneously, and during this process the distance between the sensor and the bed does not remain constant.
> 
> That's why you do the paper test to make sure that the nozzle is still at the distance it should be, then it takes readings of the eddy currents at the exact same spot to know how they change. From my understanding the sensor readings change much more drastically due to temperature than thermal expansion affects the measurements.

I'm just expressing my opinion and suggesting a way to make it easier to use. After all, doing paper calibration multiple times with the heated bed on is a painful process.

Arksine commented:
> I'm just expressing my opinion and suggesting a way to make it easier to use. After all, doing paper calibration multiple times with the heated bed on is a painful process.

I understand.  Unfortunately the manual probe is required to take valid samples.  Without the manual probe the distance between the probe's coil and the bed will change due to thermal expansion.   This would result in a calibration that is invalid.

As mentioned previously in this thread, if its possible to implement contact probing using the ldc1612 then the manual probe can be replaced with an automatic probe.  Until then, the current solution is the only workable solution for a probe mounted to the gantry.

Another alternative would be to calibrate the probe before mounting/installation.  This would require designing a printed holder that sits on the bed, with a single motor capable of moving the probe 4 to 5 mm vertically.  This would allow users to preheat the bed/chamber, then insert the probe/holder and perform the calibration.

Sineos commented:
> Another alternative would be to calibrate the probe before mounting/installation. This would require designing a printed holder that sits on the bed, with a single motor capable of moving the probe 4 to 5 mm vertically. This would allow users to preheat the bed/chamber, then insert the probe/holder and perform the calibration.

Is this so much different from what Cartographer does?
It seems to take the profile data at various heights and various temperatures via a macro

Arksine commented:
> Is this so much different from what Cartographer does?
It seems to take the profile data at various heights and various temperatures via a macro

Its possible, I haven't looked at the details of the Cartographer implementation.  Since their extension seemed to be heavily derived from the beacon extension  I assumed they performed thermal compensation in a similar fashion, where the probe is shipped with some presets in flash memory.

Sineos commented:
> I assumed they performed thermal compensation in a similar fashion

I have not seen something similar for the beacon. Details here: https://docs.cartographer3d.com/cartographer-probe/fine-tuning/temperature-differential-calibration-beta

qqqwqk commented:
> 
> As mentioned previously in this thread, if its possible to implement contact probing using the ldc1612 then the manual probe can be replaced with an automatic probe. Until then, the current solution is the only workable solution for a probe mounted to the gantry.

Maybe we can try using adxl345-probe?
https://github.com/jniebuhr/adxl345-probe
Does klipper support using multiple probes now?

Arksine commented:
> I have not seen something similar for the beacon. Details here: https://docs.cartographer3d.com/cartographer-probe/fine-tuning/temperature-differential-calibration-beta

Interesting.  That method appears to me that their method will introduce error from thermal expansion.  Personally I think performing ~20 manual probes in 30 minutes is preferable to waiting hours for a calibration procedure.  They also note that its required to re-calibrate the "model", which seems incorrect to me.  Thermal drift calibration should have no impact on the base calibration.

> Maybe we can try using adxl345-probe?
https://github.com/jniebuhr/adxl345-probe
Does klipper support using multiple probes now?

No, I don't believe that klipper supports multiple probes.  I was referring to the possibility of using the ldc1612 to detect contact.  The beacon apparently supports this method of probing.  Presumably such an implementation would support both a contact and a non-contact probing mode.

One other possibility would be to temporarily fix a switch to the tool that triggers just before nozzle contact.  It could be configured as a `z endstop`, and the module could replace the manual probe with "z home".  We would need to know its offsets, so it would effectively be a temporary secondary probe.

qqqwqk commented:
> 
> Interesting. It appears to me that method will introduce error from thermal expansion. Personally I think performing ~20 manual probes in 30 minutes is preferable to waiting hours for a calibration procedure. They also note that its required to re-calibrate the "model", which seems incorrect to me. Thermal drift calibration should have no impact on the base calibration.

In fact, if you read the installation process rigorous
(https://docs.cartographer3d.com/cartographer-probe/installation-and-setup/v2-and-v3-probes-w-input-shaping/cartographer-with-input-shaping-v2-and-v3-hybrid),
 you will find that this detection is actually achieved by using the accelerometer's tap detection function( 
they named it "tap-probe" 
https://github.com/Cartographer3D/cartographer-klipper/blob/master/idm.py#L176). 
Unfortunately, this function has not been merged into the master branch of klipper.

Sineos commented:
> Personally I think performing ~20 manual probes in 30 minutes is preferable to waiting hours for a calibration procedure.

My point is less the comfort argument. Doing 20 manual probes once does not seem too bad.
A while back, I did a small measuring campaign on manual probing and I really tried hard, but the accuracy and repeatability of the manual probes have been really bad. 
Actually, my goal was to compare the "paper method" with using a feeler gauge and IIRC, the paper method was even better than the feeler gauge. Not sure that I kept the data somewhere, though.

Of course, I cannot judge the impact of this measuring error in the probe's calibration.

> One other possibility would be to temporarily fix a switch to the tool that triggers just before nozzle contact. It could be configured as a `z endstop`, and the module could replace the manual probe with "z home". We would need to know its offsets, so it would effectively be a temporary secondary probe.

The relative position between switch and nozzle would be affected by thermal expansion as well. Wouldn't this not just introduce another kind of error?

qqqwqk commented:
I just installed voron-tap and btt-eddy on the toolhead 
so maybe we can make some changes

aaronevers commented:
If the issue is coil drift with coil temperature changes, why heat the chamber at all?  Seems like you could eliminate a lot of error if you could heat the coil directly and measure the drift without moving the toolhead at all.  Perhaps a future version could include a small heater inside the Eddy housing.
